p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/ProjectDependencyClosureGraph.java

```diff
@@ -55,88 +71,166 @@ class ProjectDependencyClosureGraph implements ProjectDependencyClosure {
+	private Map<MavenProject, Set<Edge>> projectEdgesMap = new HashMap<>();
+
+	ProjectDependencyClosureGraph(Map<MavenProject, Collection<IInstallableUnit>> projectIUMap) throws CoreException {
 		this.projectIUMap = projectIUMap;
-		this.slicer = slicer;
-		Collection<IInstallableUnit> availableIUs = projectIUMap.values().stream().flatMap(Collection::stream)
-				.collect(Collectors.toSet());
-		projectDependenciesMap = computeProjectDependencies(
-				projectIUMap.keySet(), new CollectionResult<>(availableIUs), projectIUMap);
+		// Build IU to project mapping
 		for (var entry : projectIUMap.entrySet()) {
 			MavenProject mavenProject = entry.getKey();
 			for (IInstallableUnit iu : entry.getValue()) {
 				iuProjectMap.put(iu, mavenProject);
 			}
 		}
+		// Build the graph structure
+		buildGraph();
+		// Compute project dependencies
+		projectDependenciesMap = computeProjectDependenciesFromGraph();
+		// Write DOT file if requested
+		if (DUMP_DATA) {
+			try {
+				File dotFile = new File("project-dependencies.dot");
+				dump(dotFile);
+			} catch (IOException e) {
+				// Ignore dump errors
+			}
+		}
 	}
 
 	/**
-	 * Given a set of projects, compute the mapping of a project to its dependencies
-	 * 
-	 * @param projects    the projects to investigate
-	 * @param avaiableIUs all available units that should be used to fulfill project
-	 *                    requirements
-	 * @return a Map from the passed projects to their dependencies
-	 * @throws CoreException if computation failed
+	 * Build the internal graph structure based on projectIUMap.
+	 * For each project, we collect all requirements from all IUs provided by that project,
+	 * then create edges to the capabilities that satisfy those requirements.
 	 */
-	private Map<MavenProject, ProjectDependencies> computeProjectDependencies(Collection<MavenProject> projects,
-			IQueryable<IInstallableUnit> avaiableIUs, Map<MavenProject, Collection<IInstallableUnit>> projectIUMap)
-			throws CoreException {
-		List<CoreException> errors = new CopyOnWriteArrayList<>();
-		Map<MavenProject, ProjectDependencies> result = new ConcurrentHashMap<>();
-		projects.parallelStream().unordered().takeWhile(nil -> errors.isEmpty()).forEach(project -> {
-			try {
-				ProjectDependencies projectDependencies = computeProjectDependencies(
-						Set.copyOf(projectIUMap.get(project)), avaiableIUs);
-				result.put(project, projectDependencies);
-				if (DUMP_DATA) {
-					File file = new File(project.getBasedir(), "project-dependencies.xml");
-					try {
-						new MetadataIO().writeXML(
-								Collections.unmodifiableCollection(projectDependencies.getDependencies(List.of())),
-								file);
-					} catch (IOException e) {
+	private void buildGraph() {
+		// Collect all IUs from all projects for capability matching
+		List<IInstallableUnit> allIUs = projectIUMap.values().stream()
+				.flatMap(Collection::stream)
+				.collect(Collectors.toList());
```

Comment:
We should flatmap IUs into Capabilties here first to prevent multiple objects being created.